scripts/machine_translation/train_transformer.py

```diff
@@ -0,0 +1,448 @@
+    cfg.freeze()
+    model = TransformerNMTModel.from_cfg(cfg)
+    model.initialize(mx.init.Xavier(magnitude=args.magnitude),
+                     ctx=ctx_l)
+    model.hybridize()
+    logging.info(model)
+    label_smooth_loss = LabelSmoothCrossEntropyLoss(num_labels=len(tgt_vocab),
+                                                    alpha=args.label_smooth_alpha,
+                                                    from_logits=False)
+    label_smooth_loss.hybridize()
+    rescale_loss = 100.0
+    # Construct the trainer
+    # TODO(sxjscience) Support AMP
+    if args.lr is None:
+        base_lr = 2.0 / math.sqrt(args.num_units) / math.sqrt(args.warmup_steps)
+    else:
+        base_lr = args.lr
+    lr_scheduler = InverseSquareRootScheduler(warmup_steps=args.warmup_steps, base_lr=base_lr,
+                                              warmup_init_lr=args.warmup_init_lr)
+    trainer = gluon.Trainer(model.collect_params(), 'adam',
+                            {'learning_rate': args.lr, 'beta1': 0.9,
+                             'beta2': 0.98, 'epsilon': 1e-9, 'lr_scheduler': lr_scheduler})
+    # Load Data
+    if args.bucket_scheme == 'constant':
+        bucket_scheme = ConstWidthBucket()
+    elif args.bucket_scheme == 'linear':
+        bucket_scheme = LinearWidthBucket()
+    elif args.bucket_scheme == 'exp':
+        bucket_scheme = ExpWidthBucket(bucket_len_step=1.2)
+    else:
+        raise NotImplementedError
+    # TODO(sxjscience) Try to use the new DataLoader in https://github.com/apache/incubator-mxnet/pull/17841
+    batchify_fn = bf.Tuple(bf.Pad(), bf.Pad(), bf.Stack(), bf.Stack(), bf.Stack())
+    # TODO(sxjscience) Support auto-bucket-size tuning
+    train_batch_sampler = FixedBucketSampler(lengths=[(ele[2], ele[3]) for ele in data_train],
+                                             batch_size=args.batch_size,
+                                             num_buckets=args.num_buckets,
+                                             ratio=args.bucket_ratio,
+                                             shuffle=True,
+                                             use_average_length=True,
+                                             bucket_scheme=bucket_scheme,
+                                             seed=args.seed)
+    train_data_loader = gluon.data.DataLoader(data_train, batch_sampler=train_batch_sampler,
+                                              batchify_fn=batchify_fn,
+                                              num_workers=0)
+    logging.info(train_batch_sampler)
+    val_data_loader = gluon.data.DataLoader(data_val, batch_size=16, batchify_fn=batchify_fn,
+                                            num_workers=0, shuffle=False)
+    num_batches = len(train_data_loader)
+    for v in model.collect_params().values():
+        if v.grad_req != 'null':
+            v.grad_req = 'add'
+    model.collect_params().zero_grad()
+    model_averager = AverageSGDTracker(model.collect_params())
```

Comment:
Are you using a-suffix averaged sgd, i.e., only averaging over last K iterations/epochs ?, where K is a user defined hyper-parameter.